### ISSUE TITLE
Memory failure with HMC over big data analysis model

### ISSUE THREAD

**Issue by null-a** [probmods/webppl#420]
From the [mailing list](https://groups.google.com/d/msg/webppl-dev/E9ctmD83p4w/71px9qnTBgAJ):

> The model is up here https://github.com/mhtess/habituals-tj-hmc

> The program first loads some ERPs that I’ve computed beforehand…, and then attempts to do HMC. It takes about 2 minutes, but then hits this memory issue. I don’t get the memory issue with IncrementalMH. But unfortunately, IncrMH doesn’t explore the parameters of interest very well (even after running it for 50k iterations, which takes about 8 hours)

**Comment by null-a** [comment]
I think I see another problem (aside from the memory issue) with getting this model working well with HMC. Specifically, I don't think `cache` (used with RSA) will work as expected for functions which take ad nodes as arguments.

At present `cache` will try to stringify any ad valued args, which is probably not what we want. If we fix that, then I think there's a further issue: values pulled from the cache will be fragments of the ad graph which may come from previous executions. This means that when we do backprop the derivatives are going to end up at the ad nodes for random choices from some previous execution rather than the current one. I think.

**Comment by dritchie** [comment]
Hmm, yeah, I see the problem.

The easiest fix is not to cache if any of the inputs to a call are AD nodes--that'll at least be correct.

I also think it's feasible to be able to cache the forward computation. `cache` would need to do the following: the first time it sees a given set of inputs which include AD nodes it does the usual thing (compute and store the result). On subsequent times, it matches up the AD nodes in the input with the corresponding AD nodes from the *original* input (i.e. from the first time, where we actually ran the forward computation). It then sets the current input nodes as new parents of their corresponding original input nodes using an identity op, so that derivatives propagate from the original nodes to the current nodes correctly. I'm pretty sure this will work, and it should have the right behavior with zeroDerivatives. My one remaining concern with this is that it'll create a lot of 'dangling' parent nodes (read: inserting a new parent into the AD graph isn't an operation we currently support and maybe for good reason). Basically, the original input nodes will end up with one parent for every subsequent call to the cached function (in HMC, that's one for each leapfrog step)...and every one of those parent graphs will get traversed when we do backprop, which is a ton of wasted compute, since we no longer care about any of those graphs except the current one. We'd need to find some way to break that parent connection once we no longer need it, but I'm not 100% sure right now when we can safely do that. Maybe once backprop passes the identity op that's used to make the connection? It'd be like a destructive backprop operation that modifies the graph. I'm pretty jetlagged right now, so I might not be thinking the most clearly.

As far as being able to cache both the forward *and* derivative computations: this seems really hard to implement. In general, you can't know in advance which values in the cached function's output will be part of derivative computations later and need caching (it could be a complex, nested object with AD nodes somewhere in it). I can vaguely imagine a scheme where you find all the AD nodes in the output and then give those some kind of caching backprop operator...but it sounds pretty hairy to me.

**Comment by ngoodman** [comment]
> At present cache will try to stringify any ad valued args, which is probably not what we want.

yeah, good catch. i guess we want to key on the (stringified) `ad.value`?

> I think there's a further issue: values pulled from the cache will be fragments of the ad graph which may come from previous executions. This means that when we do backprop the derivatives are going to end up at the ad nodes for random choices from some previous execution rather than the current one.

as long as `cache` is applied to a deterministic computation (as in RSA using enumeration) this should be the desired behavior. i think?

**Comment by ngoodman** [comment]
actually, for the case of RSA, where we have HMC --> cache --> Enumerate -->... (ie the subcomputation is enumerated, therefore deterministic) i think we just want to cache on keys that un-ad or that ignore the continuous values, and reset the cache between HMC steps. we do actually want to share the gradient calculation of shared sub-computations.... i think reseting cache like this may also clear up the memory leak.

**Comment by null-a** [comment]
> and reset the cache between HMC steps. we do actually want to share the gradient calculation of shared sub-computations....

Indeed. One simple way I was looking at for achieving something similar was to write the model with the RSA functions nested within the BDA loop. This means that `cache` doesn't have to deal with ad nodes (because they enter by been closed over), but more importantly we avoid pulling in the ad graph from previous executions because the cache is recreated at appropriate times.

(Although the cache will stick around long enough that MH steps can make use of it. I think this is probably OK, but it I'm entirely sure.)

It looks something like this:

https://gist.github.com/null-a/188c4958719abea1c04f67c0da424be2#file-habituals-js-L38

I tried this out and it looks like it _might_ work. Previously the acceptance ratio was fixed at zero and there were `null`s everywhere from the wonky serialization, but this looks much more sensible. It is however rather slow - a single HMC step takes a few seconds.

>  i think reseting cache like this may also clear up the memory leak.

Yep, this version seems less leaky. Though it still uses a lot of memory.

**Comment by null-a** [comment]
Also, it seems like it might be useful to be able to control the number of MH steps that are taken between each HMC iteration. Doing a long HMC simulation where all the continuous variables may move a long way and then only flipping a single discrete choice afterwards seems kinda unbalanced.

**Comment by ngoodman** [comment]
> One simple way I was looking at for achieving something similar was to write the model with the RSA functions nested within the BDA loop. This means that cache doesn't have to deal with ad nodes (because they enter by been closed over), but more importantly we avoid pulling in the ad graph from previous executions because the cache is recreated at appropriate times.

This seems like a good solution for this case!

> It is however rather slow - a single HMC step takes a few seconds.

Hmm. I wonder what makes it so slow? Is it actually the cost of creating the closures (or not having them around long enough to optimize)? Or is it just a big computation? But MH steps don't take so long, so something must be going on with AD? Maybe?

We really need some (tools for) profiling webppl!

**Comment by null-a** [comment]
> Hmm. I wonder what makes it so slow?

I'm not sure, but from the little bit of tinkering I've done, it looks like initializing the chain takes about as long (~ 2 seconds) with both HMC and MH. An HMC step then takes a little bit longer still, which makes sense since there's extra work to do e.g. backprop. This is pretty reassuring - perhaps there's nothing too strange going on here.

I'd guess that the MH steps are much faster because they make use of values cached on previous executions. From the discussion so far, this sounds like it will be pretty tricky to do in general for HMC. And since we've updated all the continuous variables, there's a limited amount that could be re-used anyway.

Of course, comparing one step of HMC with one step of MH isn't all that useful. Maybe its worth spending all that time if we get good samples out of HMC.

**Comment by mhtess** [comment]
@null-a the model rewrite you linked to isn't exactly correct. in particular, `speaker1` (and hence, `listener1`) should not be `cache`d because of the continuously valued `speaker1_optimality` parameter. If I uncache those functions, I hit the memory error again

**Comment by null-a** [comment]
The idea behind writing it this way is that the RSA functions are recreated on every execution of the `foreach` body. The _intention_ is that every time new values are sampled for the continuous variables, we get new instances of the RSA functions with fresh/empty caches.

In this way we avoid returning from the cache, values computed using old values of the continuous variables.

Does that help any? If not, perhaps it would be useful if you could say why you think functions of continuous variables shouldn't be cached.

> If I uncache those functions, I hit the memory error again

Yes. I suspect that getting `cache` working correctly with HMC is going to be important in getting this type of model working with HMC. Because with out it, I think the computational graph built by ad behind the scenes is huge.

**Comment by mhtess** [comment]
That makes sense. Thanks for clearing that up

**Comment by null-a** [comment]
For reference, [this](https://github.com/thegricean/overinformativeness/blob/a76f3000fbf10c5d0c391fa1909351f5660b6fef/models/1_basic_overinformativeness/refModule/overinformative_model.wppl#L63) looks like a nice way to structure the kind of models we've discussed here to work with HMC and `cache` as currently implemented.
